org/mozilla/jss/crypto/JSSOAEPParameterSpec.c

```diff
@@ -0,0 +1,151 @@
+#include <nss.h>
+#include <pk11pub.h>
+#include <pkcs11n.h>
+#include <pkcs11t.h>
+#include <jni.h>
+
+#include "_jni/org_mozilla_jss_crypto_JSSOAEPParameterSpec.h"
+
+#include "jssutil.h"
+#include "java_ids.h"
+#include "jss_exceptions.h"
+#include "pk11util.h"
+
+#include "NativeEnclosure.h"
+#include "StaticVoidPointer.h"
+
+PRStatus
+oaep_GetHashAlg(JNIEnv *env, jobject this, jclass this_class, CK_MECHANISM_TYPE *ret)
+{
+    jfieldID field_id = NULL;
+
+    field_id = (*env)->GetFieldID(env, this_class, "hashAlg", "J");
+    if (field_id == NULL) {
+        return PR_FAILURE;
+    }
+
+    *ret = (*env)->GetLongField(env, this, field_id);
+    return PR_SUCCESS;
+}
+
+PRStatus
+oaep_GetMGFType(JNIEnv *env, jobject this, jclass this_class, CK_RSA_PKCS_MGF_TYPE *ret)
+{
+    jfieldID field_id = NULL;
+
+    field_id = (*env)->GetFieldID(env, this_class, "mgf", "J");
+    if (field_id == NULL) {
+        return PR_FAILURE;
+    }
+
+    *ret = (*env)->GetLongField(env, this, field_id);
+    return PR_SUCCESS;
+}
+
+PRStatus
+oaep_GetSpecifiedSourceData(JNIEnv *env, jobject this, jclass this_class, CK_VOID_PTR *ret, CK_ULONG *ret_len)
+{
+    jfieldID field_id = NULL;
+    jbyteArray data = NULL;
+
+    field_id = (*env)->GetFieldID(env, this_class, "sourceData", "[B");
+    if (field_id == NULL) {
+        return PR_FAILURE;
+    }
+
+    data = (*env)->GetObjectField(env, this, field_id);
+    if (data == NULL) {
+        *ret = NULL;
+        *ret_len = 0;
+        return PR_SUCCESS;
+    }
+
+    if (!JSS_FromByteArray(env, data, (uint8_t **)ret, ret_len)) {
+        return PR_FAILURE;
+    }
+
+    return PR_SUCCESS;
+}
+
+JNIEXPORT void JNICALL
+Java_org_mozilla_jss_crypto_JSSOAEPParameterSpec_acquireNativeResources(JNIEnv *env, jobject this)
+{
+    jclass this_class = NULL;
+
+    CK_MECHANISM_TYPE hashAlg;
+    CK_RSA_PKCS_MGF_TYPE mgf;
+    CK_RSA_PKCS_OAEP_SOURCE_TYPE source = CKZ_DATA_SPECIFIED;
+    CK_VOID_PTR pSourceData = NULL;
+    CK_ULONG ulSourceDataLen = 0;
+    CK_RSA_PKCS_OAEP_PARAMS_PTR oaep_params = NULL;
+
+    jobject params_obj;
+
+    this_class = (*env)->GetObjectClass(env, this);
+    if (this_class == NULL) {
+        return;
+    }
+
+    if (oaep_GetHashAlg(env, this, this_class, &hashAlg) != PR_SUCCESS) {
+        goto failure;
+    }
+
+    if (oaep_GetMGFType(env, this, this_class, &mgf) != PR_SUCCESS) {
+        goto failure;
```

Comment:
Same thing here.

Reply:
ACK, I've added a comment after this one mentioning allocated data.